gt/main.c

```diff
@@ -1409,6 +1409,54 @@ gt_process_unparsed_incoming_pkt(struct acl_search *acl4,
 	rte_pktmbuf_free(pkt);
 }
 
+static void
+return_message(struct gt_instance *instance)
+{
+	int ret;
+	unsigned lcore_id = rte_lcore_id();
+	size_t reply_len;
+	struct dynamic_config *dy_conf = get_dy_conf();
+	struct dy_cmd_entry *entry;
+	const char *reply_msg = lua_tolstring(instance->lua_state, -1, &reply_len);
+	if (reply_msg == NULL) {
+		GT_LOG(WARNING, "gt: new lua update returned a NULL message at lcore %u\n",
+			lcore_id);
+		goto out;
+	}
+
+	entry = mb_alloc_entry(&dy_conf->mb);
+	if (entry == NULL) {
+		GT_LOG(ERR, "gt: failed to send new lua update return to Dynamic config block at lcore %d\n",
+			dy_conf->lcore_id);
+		goto out;
+	}
+
+	if (reply_len >= RETURN_MSG_MAX_LEN) {
+		rte_memcpy(entry->u.gt.return_msg, reply_msg, RETURN_MSG_MAX_LEN);
+		entry->u.gt.return_msg[RETURN_MSG_MAX_LEN - 1] = '\0';
+		entry->u.gt.length = RETURN_MSG_MAX_LEN - 1;
+
+		GT_LOG(WARNING,
+			"gt: the return message length (%lu) exceeds the limit (%d) at lcore %u\n",
+			reply_len, RETURN_MSG_MAX_LEN, lcore_id);
+	} else {
+		rte_memcpy(entry->u.gt.return_msg, reply_msg, reply_len);
+		entry->u.gt.length = reply_len;
+	}
```

Comment:
1. The `if` test should be `reply_len > RETURN_MSG_MAX_LEN`.
2. Issue the log entry, and have `reply_len = RETURN_MSG_MAX_LEN`.
3. There's no need for lines `rte_memcpy(entry->u.gt.return_msg, reply_msg, RETURN_MSG_MAX_LEN);`, `entry->u.gt.return_msg[RETURN_MSG_MAX_LEN - 1] = '\0';` and `entry->u.gt.length = RETURN_MSG_MAX_LEN - 1;`.
4. The `else` clause should always execute after the `if` statement.

Reply:
1.a. The test should be `unlikely()`.